src/osd/OSD.cc

```diff
@@ -6418,6 +6425,37 @@ void OSD::do_command(Connection *con, ceph_tid_t tid, vector<string>& cmd, buffe
     ss << "compacted omap in " << duration << " seconds";
   }
 
+  else if (prefix == "smart") {
+    dout(1) << "smartctl is called" << dendl;
+    set<string> devnames;
+    store->get_devices(&devnames);
+    uint64_t smart_timeout = cct->_conf->get_val<uint64_t>("osd_smart_report_timeout");
+    std::string result;
+    
+    json_spirit::mObject json_map; // == typedef std::map<std::string, mValue> mObject;
+    json_spirit::mValue smart_json;
+
+    for (auto it = devnames.begin(); it != devnames.end(); ++it) { 
+      // smartctl works only on physical devices; filter out any non-physical device
+      if (it->find("dm-") != std::string::npos)
+	continue;
+
+      if (probe_smart(("/dev/" + (*it)).c_str(), smart_timeout, &result)) {
+	derr << "probe_smart failed for /dev/" << *it << ", continuing to next device"<< dendl;
+	continue;
+      }
+      ss << "smartctl output: " << result; // TODO: remove when json is ready?
+      // TODO: change to read_or_throw?
+      if (!json_spirit::read(result, smart_json)) {
+	derr << "smartctl JSON output of /dev/" + *it + " is invalid" << dendl;
+      }
+      else { //json is valid, assigning 
+	json_map[*it] = smart_json;
+      }
+      // no need to result.clear() or clear smart_json
+    }
```

Comment:
here there will eventually be a loop over json_map that assembles the final result, right?

Reply:
right, I just couldn't test it so far (due to the vm), but will do on mira.